core/dbt/context/providers.py

```diff
@@ -500,19 +500,22 @@ def resolve(
                 target_version=target_version,
                 disabled=isinstance(target_model, Disabled),
             )
-        elif (
-            target_model.resource_type == NodeType.Model
-            and target_model.access == AccessType.Private
-            # don't raise this reference error for ad hoc 'preview' queries
-            and self.model.resource_type != NodeType.SqlOperation
-            and self.model.resource_type != NodeType.RPCCall  # TODO: rm
+        elif self.manifest.is_invalid_private_ref(self.model, target_model):
+            raise DbtReferenceError(
+                unique_id=self.model.unique_id,
+                ref_unique_id=target_model.unique_id,
+                access=AccessType.Private,
+                scope=cast_to_str(target_model.group),
+            )
+        elif self.manifest.is_invalid_protected_ref(
```

Comment:
As discussed in Slack: if `restrict_access: True` for this package, let's also prevent references to `private` models.

So the example in https://github.com/dbt-labs/dbt-core/issues/7713#issuecomment-1607292067 _should_ work if `restrict_access: False` for `mypkg`, and _shouldn't_ work (raise an error) if `restrict_access: True`